src/main/java/org/xbill/DNS/Client.java

```diff
@@ -59,17 +68,26 @@ private static void runSelector() {
           timeoutTasks.forEach(Runnable::run);
         }
 
-        processReadyKeys();
+        if (run) {
+          processReadyKeys();
+        }
       } catch (IOException e) {
         log.error("A selection operation failed", e);
+      } catch (ClosedSelectorException e) {
+        // ignore
```

Comment:
The only possible way to get a `ClosedSelectorException` is when `selector.close()` has been called. The only time this is done is in `close()` in the shutdown hook, which by this time has `run` already set to false and the loop is never retried.
The only reason to catch this at all is to prevent throwing when a key is still being processed during shutdown.